*** ../bash-3.0/braces.c	Thu Dec  4 11:09:52 2003
--- braces.c	Wed Aug  4 14:34:33 2004
***************
*** 341,346 ****
    if (lhs_t == ST_CHAR)
      {
!       lhs_v = lhs[0];
!       rhs_v = rhs[0];
      }
    else
--- 341,346 ----
    if (lhs_t == ST_CHAR)
      {
!       lhs_v = (unsigned char)lhs[0];
!       rhs_v = (unsigned char)rhs[0];
      }
    else
***************
*** 403,406 ****
--- 403,407 ----
  	  pass_next = 1;
  	  i++;
+ 	  level++;
  	  continue;
  	}
