*** ../bash-3.0-patched/general.c	Wed Apr 14 23:20:13 2004
--- general.c	Wed Oct 20 16:59:59 2004
***************
*** 268,272 ****
  
  #if defined (ARRAY_VARS)
!   if ((legal_variable_starter (c) == 0) && (flags && c != '[')) /* ] */
  #else
    if (legal_variable_starter (c) == 0)
--- 268,272 ----
  
  #if defined (ARRAY_VARS)
!   if ((legal_variable_starter (c) == 0) && (flags == 0 || c != '[')) /* ] */
  #else
    if (legal_variable_starter (c) == 0)
