*** ../bash-3.0/jobs.c	Fri Apr 23 16:28:25 2004
--- jobs.c	Wed Aug 18 11:15:07 2004
***************
*** 998,1002 ****
  
    if (job != NO_JOB)
!     printf ("[%d] %ld\n", job + 1, (long)pid);
    else
      programming_error (_("describe_pid: %ld: no such pid"), (long)pid);
--- 998,1002 ----
  
    if (job != NO_JOB)
!     fprintf (stderr, "[%d] %ld\n", job + 1, (long)pid);
    else
      programming_error (_("describe_pid: %ld: no such pid"), (long)pid);
***************
*** 1779,1784 ****
      {
        fail = 0;
!       for (p = jobs[job]->pipe; p->next != jobs[job]->pipe; p = p->next)
!         if (p->status != EXECUTION_SUCCESS) fail = p->status;
        return fail;
      }
--- 1779,1789 ----
      {
        fail = 0;
!       p = jobs[job]->pipe;
!       do
! 	{
! 	  if (p->status != EXECUTION_SUCCESS) fail = p->status;
! 	  p = p->next;
! 	}
!       while (p != jobs[job]->pipe);
        return fail;
      }
***************
*** 2312,2321 ****
  
    if (foreground == 0)
!     fprintf (stderr, "[%d]%c ", job + 1,
  	   (job == current_job) ? '+': ((job == previous_job) ? '-' : ' '));
  
    do
      {
!       fprintf (stderr, "%s%s",
  	       p->command ? p->command : "",
  	       p->next != jobs[job]->pipe? " | " : "");
--- 2317,2326 ----
  
    if (foreground == 0)
!     printf ("[%d]%c ", job + 1,
  	   (job == current_job) ? '+': ((job == previous_job) ? '-' : ' '));
  
    do
      {
!       printf ("%s%s",
  	       p->command ? p->command : "",
  	       p->next != jobs[job]->pipe? " | " : "");
***************
*** 2325,2334 ****
  
    if (foreground == 0)
!     fprintf (stderr, " &");
  
    if (strcmp (wd, jobs[job]->wd) != 0)
!     fprintf (stderr, "	(wd: %s)", polite_directory_format (jobs[job]->wd));
  
!   fprintf (stderr, "\n");
  
    /* Run the job. */
--- 2330,2339 ----
  
    if (foreground == 0)
!     printf (" &");
  
    if (strcmp (wd, jobs[job]->wd) != 0)
!     printf ("	(wd: %s)", polite_directory_format (jobs[job]->wd));
  
!   printf ("\n");
  
    /* Run the job. */
